### PR TITLE
Feat kv commands

### LINKED ISSUE #339: RFC - KV Commands / Queries
# Overview

Providing a fully CLI-driven experience for KV requires that users interact with the KV API in order to perform operations on and query namespaces and key value pairs for their account outside the context of a Worker project. Using Wrangler it should be possible for a developer to fully automate deploying and maintaining a Workers/KV project, complete with provisioning namespaces, writing key/value pairs, and pushing code (if necessary). This means providing basic commands for interacting with all of the documented KV endpoints:

- Create namespace #340 ✅ 
- Remove namespace #341 ✅ 
- List all namespaces associated with your account #342 ✅ 
- Rename namespace #348 ✅ 
- Write value for key in a namespace (with or without expiration) #343 
- Delete key value pair in a namespace #344 
- List all keys for a namespace #345 
- Read the value for a key within a namespace #349
- bulk write/delete for key/value pairs. #346 #347 (needs design)

## Interactions

``` sh
$ wrangler kv:namespace create <title> [--env]
$ wrangler kv:namespace delete <binding> [--env]
$ wrangler kv:namespace rename <binding> <new-title> [--env]
$ wrangler kv:namespace list [--env]
$ wrangler kv:key write <binding> <key>=<value> [--env | --ttl | --exp]
$ wrangler kv:key delete <binding> <key> [--env]
$ wrangler kv:key list <binding> [--env]
$ wrangler kv:bulk write <binding> ./path/to/data.json [--env]
$ wrangler kv:bulk delete <binding> ./path/to/data.json [--env]
```

~Stretch goal: bulk upload/delete~ these are required.

## Pre-work - not necessary before, will be post-work.

#63 would be a nice-to-have prior to this, but most of the work is in the http module right now so a big refactor is not necessary. Perhaps wrapping that and providing a `kv_client` with methods for each action would be appropriate. If that pattern is appreciated we can extend it to the existing workers calls with a `workers_client` and a `preview_client`.

## Security

One security consideration: we should empower users to protect sensitive data. Leaving values as plain text arguments in the command will leak data into shell history wherever the data is written; we need to take this into consideration either in documentation, or in the design of the command.

### PR DESCRIPTION
**Pull request by ashleymichal**
Closes #339. Comprises complete support for KV commands in Wrangler, should close https://github.com/cloudflare/wrangler/milestone/7 also.

keep this branch up to date with master using ui merges rather than rebasing.

### CODE REVIEW

File: src/commands/kv/mod.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    let user = settings::global_user::GlobalUser::new()?;
+
+    Ok(HTTPAPIClient::new(Credentials::from(user)))
+}
+
+fn account_id() -> Result<String, failure::Error> {
+    let project = settings::project::Project::new()?;
+    // we need to be certain that account id is present to make kv calls
+    if project.account_id.is_empty() {
+        panic!("Your wrangler.toml is missing the account_id field which is required to create KV namespaces!");
```
Comment (ashleygwilliams) [inline src/commands/kv/mod.rs]: any reason to panic here and not just return an error?

File: src/commands/kv/mod.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                        message::help(suggestion);
+                    }
+                }
+            }
+            APIFailure::Invalid(reqwest_err) => message::warn(&format!("Error: {}", reqwest_err)),
+        },
+    }
+}
+
+fn help(error_code: u16) -> &'static str {
```
Comment (ashleygwilliams) [inline src/commands/kv/mod.rs]: this is awesome

File: src/commands/kv/mod.rs (hunk 1 above)
Comment (ashleymichal) [reply src/commands/kv/mod.rs]: nope. should clean that up.

File: src/terminal/emoji.rs [hunk 3]
```diff
@@ -18,10 +18,11 @@ pub static DANCERS: Emoji = Emoji("👯 ", "");
 pub static EYES: Emoji = Emoji("👀 ", "");
 pub static INBOX: Emoji = Emoji("📥 ", "");
 pub static INFO: Emoji = Emoji("💁‍ ", "");
+pub static KV: Emoji = Emoji("🗂️ ", "");
 pub static MICROSCOPE: Emoji = Emoji("🔬 ", "");
 pub static SHEEP: Emoji = Emoji("🐑 ", "");
-pub static SLEUTH: Emoji = Emoji("🕵️‍♂️", "");
-pub static SPARKLES: Emoji = Emoji("✨  ", "");
+pub static SLEUTH: Emoji = Emoji("🕵️‍♂️ ", "");
```
Comment (EverlastingBugstopper) [inline src/terminal/emoji.rs]: good eye 🕵️‍♀️

File: src/commands/kv/mod.rs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        panic!("Your wrangler.toml is missing the account_id field which is required to create KV namespaces!");
+    }
+    Ok(project.account_id)
+}
+
+fn print_error(e: APIFailure) {
+    match e {
+        APIFailure::Error(_status, errors) => {
+            for error in errors {
+                message::warn(&format!("Error {}: {}", error.code, error.message,));
```
Comment (gabbifish) [inline src/commands/kv/mod.rs]: nit: any need for trailing comma here?

File: src/commands/kv/mod.rs (hunk 2 above)
Comment (gabbifish) [reply src/commands/kv/mod.rs]: YESS!!!

File: src/commands/kv/delete_namespace.rs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+use cloudflare::apiclient::APIClient;
+
+use cloudflare::workerskv::remove_namespace::RemoveNamespace;
```
Comment (gabbifish) [inline src/commands/kv/delete_namespace.rs]: should this be delete_namespace and DeleteNamespace for consistency?

File: README.md [hunk 6]
```diff
@@ -95,6 +95,9 @@ cargo install wrangler
     wrangler preview get // this is the default
     ```
 
+  - ### 🗂️ `kv`
```
[the remark this answers is not in the document]
Comment (steveklabnik) [reply README.md]: It's a card index divider, like in libraries before computers :)

File: docs/content/kv_commands.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
+# 🗂️ `kv`
+
+## Overview
+
+The `kv` subcommand allows you to store application data in the Cloudflare network to be accessed from Workers. KV operations are scoped to your account, so in order to use any of these commands, you need to:
+
+* have a Wrangler project set up with your `account_id` configured in the `wrangler.toml`
```
[the remark this answers is not in the document]
Comment (steveklabnik) [reply docs/content/kv_commands.md]: It'd be better to mention that it requires a Workers Basic subscription, if we're talking about that kind of thing in these docs.

**Comment by ashleymichal** [comment]
> Looks good so far, with some minor comments! A question: how are the endpoints for the workers kv api represented? Are they all baked into the api rust crate that areg and other folks worked on? If so, comments stating so would be super useful--I was surprised when I didn't see any explicit endpoints in this PR!

yah, the cloudflare-rs crate defines all the endpoints, which is super nice. Long term I intend to try and use it for all v4 API endpoints in wrangler, so I think commenting every endpoint would become tedious; maybe should add a section to the development contributing guide?

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (gabbifish) [reply README.md]: In Ubuntu, I swear it looks like a taco! 
Screenshot from 2019-08-13 09-48-11

File: src/commands/kv/delete_namespace.rs (hunk 5 above)
Comment (ashleymichal) [reply src/commands/kv/delete_namespace.rs]: these come from cloudflare-rs, which i believe based its naming off of the api docs. i don't think it's necessary to make that upstream change...

File: src/commands/kv/mod.rs (hunk 1 above)
Comment (ashleymichal) [reply src/commands/kv/mod.rs]: #406

File: src/commands/kv/mod.rs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        10010 | 10011 | 10012 | 10013 | 10014 | 10018 => {
+            "Run `wrangler kv list` to see your existing namespaces with IDs"
+        }
+        10009 => "Run `wrangler kv list <namespaceID>` to see your existing keys", // key errors
+        // TODO: link to more info
+        // limit errors
+        10022 | 10024 | 10030 => "See documentation",
+        // TODO: link to tool for this?
+        // legacy namespace errors
+        10021 | 10035 | 10038 => "Consider moving this namespace",
```
Comment (Electroid) [inline src/commands/kv/mod.rs]: I think this needs to be more informative: `Your namespace is running on an outdated version, consider creating a new one or using our migration tool here: ...`

Comment (ashleymichal) [reply src/commands/kv/mod.rs]: the responses from the api give a little more context, and are printed before the messages here. these just give the additional action the user can take to address the errors returned.

**Comment by ashleymichal** [comment]
Current status: All commands have been accounted for with an initial implementation. Next steps involve improving the design of the commands themselves based on stakeholder and user feedback.

File: src/commands/kv/mod.rs [hunk 9]
```diff
@@ -76,7 +76,7 @@ fn help(error_code: u16) -> &'static str {
         // legacy namespace errors
         10021 | 10035 | 10038 => "Consider moving this namespace",
         // cloudflare account errors
-        10017 | 10026 => "Check your account settings in the Cloudflare dashboard",
+        10017 | 10026 => "Workers KV is only available on the Workers Unlimited plan, please upgrade your account to use",
```
Comment (EverlastingBugstopper) [inline src/commands/kv/mod.rs]: `please upgrade your account to use` -> I feel like we need an easier call to action here, maybe a link?

Comment (EverlastingBugstopper) [reply src/commands/kv/mod.rs]: also i think i would eliminate `to use` but that's just me

File: src/commands/kv/delete_bulk.rs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+fn parse_directory(directory: &Path) -> Result<Vec<String>, failure::Error> {
+    let mut delete_vec: Vec<String> = Vec::new();
+    for entry in WalkDir::new(directory) {
+        let entry = entry.unwrap();
+        let path = entry.path();
+        if path.is_file() {
+            let key = super::generate_key(path, directory)?;
+
+            message::working(&format!("Deleting {}...", key.clone()));
```
[the remark this answers is not in the document]
Comment (gabbifish) [reply src/commands/kv/delete_bulk.rs]: This comment should definitely be more truthful... how about `"Preparing to delete {}?"`

File: src/commands/kv/list_namespaces.rs [hunk 11]
```diff
@@ -0,0 +1,41 @@
+use cloudflare::endpoints::workerskv::list_namespaces::ListNamespaces;
+use cloudflare::endpoints::workerskv::WorkersKvNamespace;
+use cloudflare::framework::apiclient::ApiClient;
+
+use prettytable::{Cell, Row, Table};
+
+use crate::terminal::message;
+
+pub fn list_namespaces() -> Result<(), failure::Error> {
+    let client = super::api_client()?;
```
[the remark this answers is not in the document]
Comment (gabbifish) [reply src/commands/kv/list_namespaces.rs]: Fixed in https://github.com/cloudflare/wrangler/pull/507 :)

File: src/commands/kv/mod.rs [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+pub use delete_key::delete_key;
+pub use delete_namespace::delete_namespace;
+pub use list_keys::list_keys;
+pub use list_namespaces::list_namespaces;
+pub use read_key::read_key;
+pub use rename_namespace::rename_namespace;
+pub use write_bulk::write_json;
+pub use write_key::write_key;
+
+// Truncate all "yes", "no" responses for itneractive delete prompt to just "y" or "n".
```
Comment (gabbifish) [inline src/commands/kv/mod.rs]: typo 
```suggestion
// Truncate all "yes", "no" responses for interactive delete prompt to just "y" or "n".
```

File: src/commands/kv/list_keys.rs [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+// Returns Some(cursor) if cursor is non-empty, otherwise returns None.
+fn get_cursor_from_result_info(result_info: Option<JsonValue>) -> Option<String> {
+    let result_info = result_info.unwrap();
+    let returned_cursor_value = &result_info["cursor"];
+    let returned_cursor = returned_cursor_value.as_str().unwrap().to_string();
+    if returned_cursor.is_empty() {
+        None
+    } else {
+        Some(returned_cursor)
+    }
```
[the remark this answers is not in the document]
Comment (gabbifish) [reply src/commands/kv/list_keys.rs]: Yes, the empty string case should be the none case! When no cursor is left, `""` is always returned as the cursor value, so calling `is_empty()` should be sufficient here (unless I am missing something else)...

File: README.md [hunk 14]
```diff
@@ -46,7 +46,15 @@ General documentation surrounding workers development and using `wrangler` can b
   wrangler generate <name> <template> --type=["webpack", "javascript", "rust"]
   ```
 
+<<<<<<< HEAD
```
Comment (EverlastingBugstopper) [inline README.md]: resolve conflict in readme

Comment (ashleymichal) [reply README.md]: nooooo

File: README.md [hunk 15]
```diff
@@ -111,13 +120,39 @@ If you would like to publish to your own domain, you will need to specify these
...
+| Key             | Value                                                                  | Example                                        |
+| --------------- | ---------------------------------------------------------------------- | ---------------------------------------------- |
+| workers_dev | false                                                                  | `workers_dev = false`                      |
+| route           | The route you would like to publish to                                 | `route = "example.com/my-worker/*"`            |
+| zone_id         | Your Cloudflare zone ID, this can be found in the Cloudflare dashboard | `zone_id = "b6558acaf2b4cad1f2b51c5236a6b972"` |
+
+>>>>>>> master
 #### Publishing the same code to multiple places
 
 If you would like to be able to publish your code to multiple places, please see the documentation for [environments](https://github.com/cloudflare/wrangler/blob/master/docs/environments.md).
```
Comment (EverlastingBugstopper) [inline README.md]: all instances of docs/environments.md need to be changed to docs/content/environments.md

**Pull request merged**